### ISSUE TITLE
TextEncoder when using V8

### ISSUE THREAD

**Issue by sashakiwi** [frida/frida#1310]
How can I get access to TextEncoder (or another fast way to convert strings or json into a byte array) when using V8?   The reason I ask is that I'm having to send sometimes large amounts of data very often and quickly and I'm finding it can get (I'm assuming) backed up causing the heap to expand - eventually killing the target process.  Various send batching strategies help somewhat, but not enough.  Digging into the send code, I see that send does a memdup on the 1st argument (after stringifying it) versus just adding a reference to the 2nd argument.  So, my theory to help this was to convert my large json into a byte array and use the 2nd argument of send for it.

Please advise if there is a way to get TextEncoder or another available way to efficiently (loop with charcodeat is too slow) convert my json to a byte array in the provided V8.  I am compiling from source if that helps.

**Comment by sashakiwi** [comment]
To add a little context:  Here is the code I'm talking about which does a g_strdup on the 1st argument versus a g_bytes_ref on the 2nd argument.
static void
gum_v8_script_emit (GumV8Script * self,
          const gchar * message,
          GBytes * data)
{
 auto d = g_slice_new (GumEmitData);
 d->script = self;
 g_object_ref (self);
 d->message = g_strdup (message);
 d->data = (data != NULL) ? g_bytes_ref (data) : NULL;
 auto source = g_idle_source_new ();
 g_source_set_callback (source, (GSourceFunc) gum_v8_script_do_emit, d,
   (GDestroyNotify) gum_v8_emit_data_free);
 g_source_attach (source, self->main_context);
 g_source_unref (source);
}


In our javascript, we have a very large arrays of JSON (items) that we are passing into send:
send({'items': items});

If we use Duktape, we could use TextEncoder to create a byte array and send it like this:
send('myItems', items_byte_array);
If we are using V8, TextEncoder is not found. 

So, I'm trying to see if we can use TextEncoder (or another efficient method to convert to byte array) somehow in our case where we use V8 OR see if there's another way to optimize heap usage in the send.

**Comment by oleavr** [comment]
That's a great question.

Quick braindump:
- Using `send()` for delivering bulk data requires implementing back pressure and limiting the message size to something like 4 MB. (As I suspect the underlying GDBus machinery in GLib may not be zero-copy all the way through, though it would be great to dig into that.) You might find [frida-remote-stream](https://github.com/nowsecure/frida-remote-stream) helpful in case you're using [frida-compile](https://github.com/oleavr/frida-agent-example) so modules from npm can be consumed by your agent.
- One potential solution would be allowing a JS string to be passed as the second argument to `send()`, so you can `JSON.stringify()` it.
- Another option is using `Socket.connect()` or `Socket.listen()`, which allow you to `writeMemoryRegion()` to write directly from memory to a TCP or UNIX socket. (That still means you'd need to `Memory.allocUtf8String(JSON.stringify(items))` – but we could improve `writeAll()` to support passing a string.
- You might be able to use off-the-shelf modules from npm to implement efficient binary serialization e.g. to protobuf, which would fit right into the current `send()` API (as Node.js `Buffer` objects' underlying `ArrayBuffer` can be accessed and passed to `send()`. This is because frida-compile uses browserify under the hood.).

**Comment by ballen127** [comment]
Hi @oleavr 

Regarding your 4th option, can you please me guide how to import off-the-shelf node modules?
I tried like this but it happened no such module error.
```
...
cateyes_script_options_set_name (options, "example");
cateyes_script_options_set_runtime (options, CATEYES_SCRIPT_RUNTIME_V8);

script = cateyes_session_create_script_sync (session,
        "require('text-encoding');\n"
        "Interceptor.attach(Module.getExportByName(null, 'open'), {\n"
        "  onEnter(args) {\n"
        "    console.log('[*] open(\"' + args[0].readUtf8String() + '\")');\n"
        "  }\n"
        "});\n",
        options, NULL, &error);
g_assert (error == NULL);
...
```
How to install external library?

Thank you!

**Comment by ballen127** [comment]
Hi @oleavr

Regarding your 4th option, can you please me guide how to import off-the-shelf node modules?
I tried like this but it happened no such module error.

...
frida_script_options_set_name (options, "example");
frida_script_options_set_runtime (options, FRIDA_SCRIPT_RUNTIME_V8);

script = frida_session_create_script_sync (session,
        "require('text-encoding');\n"
        "Interceptor.attach(Module.getExportByName(null, 'open'), {\n"
        "  onEnter(args) {\n"
        "    console.log('[*] open(\"' + args[0].readUtf8String() + '\")');\n"
        "  }\n"
        "});\n",
        options, NULL, &error);
g_assert (error == NULL);
...
How to install external library?

Thank you!

**Comment by ballen127** [comment]
Hi @oleavr

Regarding your 4th option, can you please me guide how to import off-the-shelf node modules?
I tried like this but it happened no such module error.
```
...
frida_script_options_set_name (options, "example");
frida_script_options_set_runtime (options, FRIDA_SCRIPT_RUNTIME_V8);

script = frida_session_create_script_sync (session,
        "require('text-encoding');\n"
        "Interceptor.attach(Module.getExportByName(null, 'open'), {\n"
        "  onEnter(args) {\n"
        "    console.log('[*] open(\"' + args[0].readUtf8String() + '\")');\n"
        "  }\n"
        "});\n",
        options, NULL, &error);
g_assert (error == NULL);
...
```
How to install external library?

Thank you!

**Comment by oleavr** [comment]
@ballen127 It would be a separate build step that compiles it by invoking `frida-compile`. You would then load the generated .js or include it as a resource. The easiest way to get started is cloning this repo: https://github.com/oleavr/frida-agent-example – running `npm install` in it will fetch `frida-compile` and invoke the `prepare` script in package.json, which in turn invokes the `build` script, which does: `frida-compile agent/index.ts -o _agent.js`. You would either integrate this in your build system, or add the generated .js to source control. (Which might be acceptable if it doesn't change that often.)

**Comment by sashakiwi** [comment]
Thanks for the brain dump!!

How would option 2 work? That would have to be a frida code change, right?

**Comment by oleavr** [comment]
Yes, it would require a change [here](https://github.com/frida/frida-gum/blob/09c2c63ef49d8e0a0cc05896eea7ad898dff3e11/bindings/gumjs/gumv8core.cpp#L1174) (for V8) and [here](https://github.com/frida/frida-gum/blob/09c2c63ef49d8e0a0cc05896eea7ad898dff3e11/bindings/gumjs/gumdukcore.c#L1766) (for Duktape). You can try it locally and profile it, and it if it looks good I'd be happy to assist with the PR.

To try it locally, change:

```C
  if (!_gum_v8_args_parse (args, "sB?", &message, &data))
```

to:

```C
  if (!_gum_v8_args_parse (args, "sB~?", &message, &data))
```

Which means it will use `_gum_v8_bytes_parse()` instead of `_gum_v8_bytes_get()`, so it also accepts a string in addition to `ArrayBuffer | *View | number[]`.

**Comment by sashakiwi** [comment]
Thank you!  I will do some speed and memory profiling with options 2 and 4 - and report my results - to see which way to go in the short term.  Option 1 looks very interesting and may end up the long-term solution if needed, but would require more significant changes.  Thanks again for pointing that out and for constantly improving Frida.

**Comment by oleavr** [comment]
Sounds like a great plan! 👍  Would love to make `send()` faster, it hasn't been profiled a whole lot so we might discover some low-hanging fruits.

By the way, as you might already be aware, we have a Chrome Inspector/DevTools integration that could prove helpful for profiling the JS side. It's only a matter of passing `--debug` (CLI tools) or calling `session.enable_debugger()` (API) to enable it. Note that it will be listening on the host, even if the agent is on a remote system.

**Comment by sashakiwi** [comment]
While testing this, I found an interesting thing.  There seems to be extra characters added at the end of the data argument of send when I pass a string to it.   Unless I'm missing something....

Here's the javascript snippet:
           "var pending = [];\n"
              ....  pending populated.....

               "var itemList = JSON.stringify({'items':pending});\n"
               "console.log('itemListdirect ' + itemList);\n"        -> output = my_string
            "console.log(itemList.length);\n"                             -> output = correct string length
            "send('items', itemList);\n"


Here's the on_message snippet:
       void on_message (FridaScript *script, const gchar *message, GBytes *data, gpointer user_data){
            ...... 
            const gchar * str_data;
            gsize size;
            if(data != NULL)
            {
                str_data = (const gchar *)g_bytes_get_data (data, &size);
                g_print ("Got Gdata message %s\n",str_data);           -> output = my_string + random character(s)
                g_print ("Got Size %d\n",size);                                     -> output = correct string length matching javascript output
            }
              string msg1 (str_data);
             g_print("new string %s\n",msg1.c_str();                        -> output = [Invalid UTF-8] my_string + 1+ random character(s)
              string msg (str_data,size);
              g_print ("new string %s\n",msg.c_str());                       -> output = my_string (no extra character)

Regarding testing, I've tested with a heavy (running as fast as it can with no stops) single-threaded target process and see only a slight improvement in heap usage due to this change - still worth doing though.  Single-threaded target process runtime(performance impact) is the same.  I'm still going to test with multi-threaded target processes.    I would be interested to know if the option 1 (frida-remote-stream) has been heavily used/tested yet?  If so, does it (or would you expect it to) reduce the target process runtime performance impact or heap usage?

Thanks!!

**Comment by jpstotz** [comment]
@oleavr Regarding option 2 "allowing a JS string to be passed as the second argument to `send()`:

You have outlined the necessary changes for implementing option 2 in https://github.com/frida/frida/issues/1310#issuecomment-634210225. Do you expect any potential problems e.g. regarding performance or are there different reasons why this really useful improvement was never applied?
